yew-macro/src/classes/mod.rs

```diff
@@ -0,0 +1,71 @@
+use proc_macro2::TokenStream;
+use quote::{quote, ToTokens};
+use syn::parse::{Parse, ParseStream};
+use syn::punctuated::Punctuated;
+use syn::{Expr, ExprLit, Lit, LitStr, Token};
+
+/// List of HTML classes.
+pub struct Classes(Punctuated<ClassExpr, Token![,]>);
+
+impl Parse for Classes {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        input.parse_terminated(|x| ClassExpr::parse(x)).map(Self)
+    }
+}
+
+impl ToTokens for Classes {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let n = self.0.len();
+        let push_classes = self.0.iter().map(|x| match x {
+            ClassExpr::Lit(class) => quote! {
+                unsafe { __yew_classes.unchecked_push(#class) };
+            },
+            ClassExpr::Expr(class) => quote! {
+                __yew_classes.push(#class);
+            },
+        });
+        tokens.extend(quote! {
+            {
+                let mut __yew_classes = ::yew::html::Classes::with_capacity(#n);
+                #(#push_classes)*
+                __yew_classes
+            }
+        });
+    }
+}
+
+#[allow(clippy::large_enum_variant)]
+enum ClassExpr {
+    Lit(LitStr),
+    Expr(Expr),
+}
```

Comment:
Would you mind sharing your knowledge on this?

I don't know much about lower level programming. I thought allocating into the heap would save memory but using more space in the stack would be faster. cc @Yozhgoor (because I gave a wrong explanation).

Though :thinking: I think sometimes things get faster because they are smaller. The speed boost is due to shifting the work load to a faster component (the CPU). It is the case for example with filesystem compression where you spend less time reading bytes from IO and more time decompressing with the CPU which overall is faster.

Reply:
It's not always that straightforward. Without benchmarking these things it's really hard to tell which approach is actually faster in our case. And honestly, I don't know about most of the optimizations performed by the compiler (and especially not when it comes to LLVM). This isn't directly about speed though, Clippy complains because one variant (`Expr`) has a significantly bigger memory footprint than the others (`LitStr`). This is an issue because the size of an enum is determined by its biggest variant so in our case `ClassExpr::Lit` (which is probably the more common case) is penalized by a ton padding which in return can reduce performance.

If you look at what's contained in an `Expr` value you'll find that most of it is already on the heap anyway (either in the form of `Vec<_>` or `Box<_>`) so I don't think the overhead of the allocation causes a noticeable speed difference here. Parsing the `proc_macro2::TokenStream` into an `Expr` is probably the dominating factor for performance here anyway.

Reply:
What I explained initially is that this warns that the enum is taking too much memory because of the difference between the size of the variants is big. So putting the big variant into a Box saves memory when the smaller variant is used. But since we are in a macro, I suppose it's less of a problem. It's not like we are going to instantiate millions of classes during the compilation. That is why I think in this case we can safely ignore it.

But yeah it wasn't hard to fix either :confounded: